ddl/ddl_api.go

```diff
@@ -587,6 +587,22 @@ func buildColumnAndConstraint(
 	if err := setCharsetCollationFlenDecimal(colDef.Tp, chs, coll); err != nil {
 		return nil, nil, errors.Trace(err)
 	}
+	// Some column checks are left here, so varchar auto convert is performed. See also issue 30328
+	sessVars := ctx.GetSessionVars()
+	if sessVars != nil && !sessVars.SQLMode.HasStrictMode() && colDef.Tp.Tp == mysql.TypeVarchar {
```

Comment:
I wonder whether here needs a test too. If it is, where shall I add the test? (`db_xxx.go` is a bit confusing)